app/src/main/java/com/amaze/filemanager/utils/ServiceWatcherProgressAbstract.java

```diff
@@ -33,4 +50,130 @@ public void progressResumed() {
+                    break;
+                case ExtractService.ID_NOTIFICATION:
+                    titleResource = R.string.extracting;
+                    break;
+                case ZipService.ID_NOTIFICATION:
+                    titleResource = R.string.compressing;
+                    break;
+                default:
+                    titleResource = R.string.processing;
+                    break;
+            }
+
+            mBuilder.setContentTitle(context.getResources().getString(titleResource));
+            mBuilder.setContentText(fileName + " " + Formatter.formatFileSize(context, writtenSize) + "/" +
+                    Formatter.formatFileSize(context, totalSize));
+            mNotifyManager.notify(id, mBuilder.build());
+
+            if (writtenSize == totalSize || totalSize == 0) {
+                if (move && id == CopyService.NOTIFICATION_ID) {
+
+                    //mBuilder.setContentTitle(getString(R.string.move_complete));
+                    // set progress to indeterminate as deletion might still be going on from source
+                    mBuilder.setProgress(0, 0, true);
+
+                    mBuilder.setContentText(context.getResources().getString(R.string.processing));
+                    mBuilder.setOngoing(false);
+                    mBuilder.setAutoCancel(true);
+                    mNotifyManager.notify(id, mBuilder.build());
+                } else {
+                    publishCompletedResult(id);
+                }
+            }
+
+            //for processviewer
+            CopyDataParcelable intent = new CopyDataParcelable(fileName, sourceFiles, sourceProgress,
+                    totalSize, writtenSize, speed, move, isComplete);
+            putDataPackage(intent);
+            if (progressListener != null) {
+                progressListener.onUpdate(intent);
+                if (isComplete) progressListener.refresh();
+            }
+        } else publishCompletedResult(id);
+    }
+
+    private void publishCompletedResult(int id1) {
+        try {
+            mNotifyManager.cancel(id1);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public interface ProgressListener {
+        void onUpdate(CopyDataParcelable dataPackage);
+
+        void refresh();
+    }
+
+    public void setProgressListener(ServiceWatcherProgressAbstract.ProgressListener progressListener) {
+        this.progressListener = progressListener;
+    }
+
+    /**
+     * Returns the {@link #dataPackages} list which contains
+     * data to be transferred to {@link ProcessViewerFragment}
+     * Method call is synchronized so as to avoid modifying the list
+     * by {@link com.amaze.filemanager.utils.ServiceWatcherUtil#handlerThread} while {@link MainActivity#runOnUiThread(Runnable)}
+     * is executing the callbacks in {@link ProcessViewerFragment}
+     *
+     * @return
+     */
+    public synchronized CopyDataParcelable getDataPackage(int index) {
```

Comment:
Why the object synchronization?

Reply:
Because this call is not atomic

Reply:
Why is it not atomic and why is it not synchronized on the list object?

Reply:
Please test and see if there are any problems.

Reply:
Currently this seems to be working fine and also serves the fix for original problem... So can be merged.. the only problem is sftp this PR wasn't focused to fix at the time.. 
Raise an issue later if you find any problem... We need to start pushing betas soon.

Reply:
> Currently this seems to be working fine and also serves the fix for original problem... So can be merged.. 

Good code isn't just fixing the existent issues, it is also fixing new issues that you haven't found yet.

Fix [this](https://github.com/TeamAmaze/AmazeFileManager/pull/916#discussion_r161290245), [this](https://github.com/TeamAmaze/AmazeFileManager/pull/916#discussion_r161062222), [this](https://github.com/TeamAmaze/AmazeFileManager/pull/916#discussion_r161287976), [this](https://github.com/TeamAmaze/AmazeFileManager/pull/916#discussion_r161288241) and [this](https://github.com/TeamAmaze/AmazeFileManager/pull/916#discussion_r161288695) (half of those are questions) with well written, efficient and non repetitive code and I'll merge.